applications/ShapeOptimizationApplication/python_scripts/algorithm_steepest_descent.py

```diff
@@ -12,70 +12,60 @@
-                  DataLogger,
-                  OptimizationSettings ):
-
-        self.ModelPartController = ModelPartController
+    def __init__( self, OptimizationSettings, MdpaController, Analyzer, Communicator ):
+        self.OptimizationSettings = OptimizationSettings
+        self.MdpaController = MdpaController
         self.Analyzer = Analyzer
         self.Communicator = Communicator
-        self.Mapper = Mapper
-        self.DataLogger = DataLogger
-        self.OptimizationSettings = OptimizationSettings
 
-        self.OptimizationModelPart = ModelPartController.GetOptimizationModelPart()
-        self.DesignSurface = ModelPartController.GetDesignSurface()
+        self.OptimizationModelPart = MdpaController.GetOptimizationModelPart()
+        self.DesignSurface = MdpaController.GetDesignSurface()
 
         self.maxIterations = OptimizationSettings["optimization_algorithm"]["max_iterations"].GetInt() + 1
         self.projectionOnNormalsIsSpecified = OptimizationSettings["optimization_algorithm"]["project_gradients_on_surface_normals"].GetBool()
         self.onlyObjectiveId = OptimizationSettings["objectives"][0]["identifier"].GetString()
         self.dampingIsSpecified = OptimizationSettings["design_variables"]["damping"]["perform_damping"].GetBool()
 
+        import mapper_factory
+        self.Mapper = mapper_factory.CreateMapper( MdpaController, OptimizationSettings )
+        import data_logger_factory
+        self.DataLogger = data_logger_factory.CreateDataLogger( MdpaController, Communicator, OptimizationSettings )
+
         self.GeometryUtilities = GeometryUtilities( self.DesignSurface )
         self.OptimizationUtilities = OptimizationUtilities( self.DesignSurface, OptimizationSettings )
         if self.dampingIsSpecified:
-            damping_regions = self.ModelPartController.GetDampingRegions()
+            damping_regions = self.MdpaController.GetDampingRegions()
             self.DampingUtilities = DampingUtilities( self.DesignSurface, damping_regions, self.OptimizationSettings )
 
     # --------------------------------------------------------------------------
-    def execute( self ):
-        self.__initializeOptimizationLoop()
-        self.__runOptimizationLoop()
-        self.__finalizeOptimizationLoop()
-
-    # --------------------------------------------------------------------------
-    def __initializeOptimizationLoop( self ):
-        self.Analyzer.initializeBeforeOptimizationLoop()
-        self.ModelPartController.InitializeMeshController()
-        self.DataLogger.StartTimer()
+    def InitializeOptimizationLoop( self ):
+        self.Analyzer.InitializeBeforeOptimizationLoop()
+        self.MdpaController.InitializeMeshController()
         self.DataLogger.InitializeDataLogging()
 
     # --------------------------------------------------------------------------
-    def __runOptimizationLoop( self ):
+    def RunOptimizationLoop( self ):
+        from custom_timer import Timer
```

Comment:
Imports should be at the top of a file (according to PEP8)

Reply:
yep, I adjusted wherever possible. Only in the analyzer_facotry. Here we just want to import the KratosInternalAnalyzer, and with it the StructuralMechanicsApp, when needed.